test/tokengenerator.go

```diff
@@ -0,0 +1,141 @@
+func NewIdentity() *Identity {
+	return &Identity{
+		ID:       uuid.NewV4(),
+		Username: "testuser-" + uuid.NewV4().String(),
+	}
+}
+
+// TokenGenerator represents the test token and key manager.
+type TokenGenerator struct {
+	keyMap map[string]*rsa.PrivateKey
+}
+
+// NewTokenGenerator creates a new TokenGenerator.
+func NewTokenGenerator() *TokenGenerator {
+	tg := &TokenGenerator{}
+	tg.keyMap = make(map[string]*rsa.PrivateKey)
+	return tg
+}
+
+// CreateKey creates and stores a new key with the given kid.
+func (tg *TokenGenerator) CreateKey(kid string) (*rsa.PrivateKey, error) {
+	reader := rand.Reader
+	key, err := rsa.GenerateKey(reader, bitSize)
+	if err != nil {
+		return nil, err
+	}
+	tg.keyMap[kid] = key
+	return key, nil
+}
+
+// Key retrieves the key associated with the given kid.
+func (tg *TokenGenerator) Key(kid string) (*rsa.PrivateKey, error) {
+	key, ok := tg.keyMap[kid]
+	if !ok {
+		return nil, errors.New("given kid does not exist")
+	}
+	return key, nil
+}
+
+// GenerateSignedToken generates a JWT user token and signs it using the default private key
+func (tg *TokenGenerator) GenerateSignedToken(identity Identity, kid string, extraClaims ...ExtraClaim) (string, error) {
+	token := jwt.New(jwt.SigningMethodRS256)
+	token.Claims.(jwt.MapClaims)["uuid"] = identity.ID
+	token.Claims.(jwt.MapClaims)["preferred_username"] = identity.Username
+	token.Claims.(jwt.MapClaims)["sub"] = identity.ID
+	token.Claims.(jwt.MapClaims)["jti"] = uuid.NewV4().String()
+	token.Claims.(jwt.MapClaims)["session_state"] = uuid.NewV4().String()
+	token.Claims.(jwt.MapClaims)["iat"] = time.Now().Unix()
+	token.Claims.(jwt.MapClaims)["exp"] = time.Now().Unix() + 60*60*24*30
+	token.Claims.(jwt.MapClaims)["nbf"] = 0
+	token.Claims.(jwt.MapClaims)["iss"] = "codeready-toolchain"
+	token.Claims.(jwt.MapClaims)["typ"] = "Bearer"
+	token.Claims.(jwt.MapClaims)["approved"] = true
+	token.Claims.(jwt.MapClaims)["name"] = "Test User"
+	token.Claims.(jwt.MapClaims)["company"] = "Company Inc."
+	token.Claims.(jwt.MapClaims)["given_name"] = "Test"
+	token.Claims.(jwt.MapClaims)["family_name"] = "User"
+	token.Claims.(jwt.MapClaims)["email_verified"] = true
+	for _, extra := range extraClaims {
+		extra(token)
+	}
+	key, err := tg.Key(kid)
+	if err != nil {
+		return "", err
+	}
+	token.Header["kid"] = kid
+	tokenStr, err := token.SignedString(key)
+	if err != nil {
+		panic(errors.WithStack(err))
+	}
+	return tokenStr, nil
+}
+
+// GetKeyService creates a http key service and return the URL
+func (tg *TokenGenerator) GetKeyService() string {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		keySet := &WebKeySet{}
+		for kid, key := range tg.keyMap {
+			thisKey, err := jwk.New(&key.PublicKey)
+			if err != nil {
+				panic(fmt.Sprintf("fatal error adding keys to key service: %s", err.Error()))
```

Comment:
we should avoid `panic` even in tests. What about return an error instead?

Reply:
Yeah, right. Panic is better than Exit() and a 500 error response is better than panic. It's not that critical as Exit() though :)

Reply:
fixed here: afa63c1cb951b6f2df5316d1b066ae6fa01bfc67